# **CASE STUDY 2 — Penilaian Kelulusan**
> *(In Bahasa) for English version
https://github.com/ardisoegi/ai-audit-case-studies/blob/main/Case_Study_2_ENG.ipynb*

> klik **> Run All** di toolbar di atas untuk melihat semua hasil proses perintah sekaligus

# 1. Overview



## **1.1 - Konteks Dataset**

Dataset: **log nilai ujian siswa berbasis AI scoring**

Dataset ini mensimulasikan output sistem penilaian ujian otomatis yang digunakan di sebuah sekolah. Setiap baris mewakili satu siswa beserta nilai. Kolom ai_score ditentukan untuk merepresentasikan nilai yang dihasilkan oleh sistem AI, bukan nilai ujian siswa sebenarnya yang dinilai manual. Dataset dihasilkan oleh generatif AI.



> *Persiapan : Jalankan kode di bawah ini untuk memuat dataset (shortcut: ctrl + Enter)*

In [ ]:
import pandas as pd
import numpy as np
from IPython.display import display

# Log nilai ujian siswa berbasis AI scoring
data = {
    "student_id": ["S001", "S002", "S003", "S004", "S005",
                   "S006", "S007", "S008", "S009", "S010",
                   "S011", "S012", "S013", "S014", "S015",
                   "S016", "S017", "S018"],
    "name": ["Jonathan", "Aisha", "Jonas", "Zara", "Liam",
             "Fatima", "Ahmed", "Sofia", "Kai", "Elena",
             "Mateo", "Amara", "Kenichi", "Leila", "Idris",
             "Anya", "Ryan", "Omar"],
    "gender": ["Male", "Female", "Male", "Female", "Male",
               "Female", "Male", "Female", "Male", "Female",
               "Male", "Female", "Male", "Female", "Male",
               "Female", "Male", "Male"],
    "class": ["A", "B", "A", "C", "B",
              "A", "C", "B", "C", "A",
              "A", "C", "B", "A", "C",
              "B", "B", "C"],
    "ai_score": [82, 78, None, 74, 88,
                 91, 76, 110, 83, 69,
                 95, 72, 80, -10, 85,
                 77, 88, 73],
    "status": ["pass", "pass", "pass", "fail", "pass",
               "pass", "fail", "pass", "pass", "fail",
               "Pass", "fail", "pass", "fail", "pass",
               "PASS", "pass", "fail"]
}

df = pd.DataFrame(data)

display(df)


,student_id,name,gender,class,ai_score,status
0,S001,Jonathan,Male,A,82.0,pass
1,S002,Aisha,Female,B,78.0,pass
2,S003,Jonas,Male,A,NaN,pass
3,S004,Zara,Female,C,74.0,fail
4,S005,Liam,Male,B,88.0,pass
5,S006,Fatima,Female,A,91.0,pass
6,S007,Ahmed,Male,C,76.0,fail
7,S008,Sofia,Female,B,110.0,pass
8,S009,Kai,Male,C,83.0,pass
9,S010,Elena,Female,A,69.0,fail


## **1.2 - Struktur Dataset**

In [ ]:
# Cek struktur dataset
print(f"Berikut adalah tampilan header dataset")
print('='*50)
display(df.head(1))


print(f"\n\nBerikut adalah shape dataset (jumlah baris, jumlah kolom)")
print('='*50)
display(df.shape)


print(f"\n\nBerikut adalah tipe dari kolom dataset")
print('='*50)
display(df.dtypes)

Berikut adalah tampilan header dataset


,student_id,name,gender,class,ai_score,status
0,S001,Jonathan,Male,A,82.0,pass




Berikut adalah shape dataset (jumlah baris, jumlah kolom)


(18, 6)



Berikut adalah tipe dari kolom dataset


,0
student_id,object
name,object
gender,object
class,object
ai_score,float64
status,object


Dari perintah di atas dihasilkan Struktur Dataset
-  **6** Header; **student_id,	name,	gender,	class,	ai_score,	status**
-  **18** Baris, **6** Kolom
-  Tipe data object; **student_id,	name,	gender,	class,	status**
-  Tipe data float64 (desimal); **ai_score**
-  Tipe dataset; **object**

> Dataset sudah teridentifikasi dan bisa dilaksanakan audit.

## **1.3 - Ringkasan Statistik dan Kategorikal**

In [ ]:

# 1. Statistik
print(f"\nTinjauan statistik dari dataset")
print('='*50)
display(df.describe())

# 2. Kategorikal

## Mencari kolom yang datanya tidak sesuai ketentuan.
print(f"\n\nJumlah macam data masing-masing kolom")
print('='*50)
display(df.nunique())

## Ditemukan kolom 'status' memiliki data tidak standar.
print(f"\n\nIsi label dari kolom 'status'")
print('='*50)
display(np.unique(df['status'].dropna()))

## Mencari kolom yang mempunyai NaN, None, atau null.
print(f"\n\nKolom numerikal yang mempunyai data 'Nan, none, atau null'")
print('='*50)
display(df.isnull().sum())

## Dari output di atas ditemukan NaN di kolom numerik 'ai_score'
print(f"\n\nList data yang memiliki nilai 'NaN, none, atau null'")
print('='*50)
display(df[pd.to_numeric(df['ai_score'], errors='coerce').isnull()])







Tinjauan statistik dari dataset


,ai_score
count,17.000000
mean,77.117647
std,24.583740
min,-10.000000
25%,74.000000
50%,80.000000
75%,88.000000
max,110.000000




Jumlah macam data masing-masing kolom


,0
student_id,18
name,18
gender,2
class,3
ai_score,16
status,4
status_normalized,2




Isi label dari kolom 'status'


array(['PASS', 'Pass', 'fail', 'pass'], dtype=object)



Kolom numerikal yang mempunyai data 'Nan, none, atau null'


,0
student_id,0
name,0
gender,0
class,0
ai_score,1
status,0
status_normalized,0




List data yang memiliki nilai 'NaN, none, atau null'


,student_id,name,gender,class,ai_score,status,status_normalized
2,S003,Jonas,Male,A,NaN,pass,pass


Dari perintah di atas mendapatkan beberapa poin temuan
1. **Statistik**
 - Ditemukan nilai diluar rentang yang diperbolehkan yaitu nilai terendah **min: -10** dan nilai tertinggi **max: 110**. Melanggar aturan rentang nilai skala **100** (0 sampai 100).
- Standart Deviasi cenderung besar **std: 24** mengindikasikan keragaman perolehan nilai dari siswa dari rata-rata **mean: 77.11**.

2. **Kategorikal**
 - Kolom **status** menghasilkan **4 label** yang seharusnya hanya **2 label** yaitu **pass** dan **fail**, kemungkinan faktor penulisan kapitalisasi yang tidak diseragamkan.
 - Ditemukan nilai **NaN** (Not a Number) di kolom **ai_score** yang seharusnya numerik pada data siswa bernama **Jonas** id **S003**



---

# 2. Analisis



## **2.1 - Checklist Audit**

1. Apakah siswa yang memperoleh nilai diluar ketentuan mendapatkan status kelulusan yang sesuai?
2. Apakah sistem memiliki validasi input untuk filtrasi data nilai yang bukan numerik atau nilai yang tidak sesuai ketentuan penilaian skala 0 sampai 100?
3. Apakah pelabelan status siswa yang tidak konsisten atau tidak diseragamkan akan mempengaruhi hasil keputusan sistem?
4. Apakah ada indikasi bias dari hasil penilaian sistem terhadap kelompok siswa berdasarkan background gender? apakah rata-rata skor dan tingkat kelulusan terdistribusi secara proporsional dan adil?

## **2.2 - Solusi Checklist 1**





**Checklist 1 - Step 1** : Mencari nilai yang diluar skala 0-100 atau nilai non valid





In [ ]:
# Mem-filter siswa berdasar kolom 'ai_score' untuk mencari nilai yang tidak sesuai skala 0 - 100 atau non valid
df_nilai = df[
    (df['ai_score'] > 100) |
    (df['ai_score'] < 0) |
    (df['ai_score'].isnull())
]
display(df_nilai)

,student_id,name,gender,class,ai_score,status
2,S003,Jonas,Male,A,NaN,pass
7,S008,Sofia,Female,B,110.0,pass
13,S014,Leila,Female,A,-10.0,fail


### **Checklist 1 - Conclusion**

Apakah siswa yang memperoleh nilai diluar ketentuan mendapatkan status kelulusan yang sesuai?
>Tidak, karena ditemukan **3 siswa** yang mempunyai nilai diluar skala 0-100 atau nilai tidak valid :
- **S003** dengan nilai **NaN**, sistem tidak memiliki nilai untuk membuat keputusan, tapi tetap menghasilkan status **pass**
- **S008** dengan nilai **110** melebihi batas maksimal valid, sistem tetap memproses dan menghasilkan **pass**
- **S014** dengan nilai **-10** di bawah batas minimum valid, sistem tetap memproses dan menghasilkan **fail**

## **2.3 - Solusi Checklist 2**

**Checklist 2 - Step 1** : Mencari banyaknya siswa yang mendapatkan nilai diluar skala 0-100 atau nilai tidak valid

In [ ]:
# Menghitung siswa yang mempunyai nilai tidak valid
nilai_tidak_valid = df_nilai.shape[0]
print(f"Banyaknya siswa yang mempunyai nilai tidak valid : {nilai_tidak_valid}")

Banyaknya siswa yang mempunyai nilai tidak valid : 3


**Checklist 2 - Step 2** : Mencari banyaknya siswa yang mendapatkan nilai sesuai ketentuan skala 0-100

In [ ]:
nilai_valid = df.shape[0] - nilai_tidak_valid
print(f"Banyaknya siswa yang mempunyai nilai valid : {nilai_valid}")

Banyaknya siswa yang mempunyai nilai valid : 15


### **Checklist 2 - Conclusion**

Apakah sistem memiliki validasi input untuk filtrasi data nilai yang bukan numerik atau nilai yang tidak sesuai ketentuan penilaian skala 0 sampai 100?
>Tidak, karena ditemukan sebanyak **3 siswa** yang mempunyai nilai diluar skala 0-100 atau nilai tidak valid tetap mendapatkan **status kelulusan** (**pass** or **fail**). Jika validasi input tidak ada, sistem tidak bisa membedakan data yang layak diproses atau yang tidak layak, ini berarti semua keputusan yang dihasilkan sistem untuk data tidak valid tidak bisa dipercaya.


## **2.4 - Solusi Checklist 3**

**Checklist 3 - Step 1** : Menguji format label di kolom 'status'




In [ ]:
# Menghitung banyaknya kemunculan data label di kolom 'status'
display(df['status'].value_counts())


,count
status,
pass,10
fail,6
Pass,1
PASS,1


**Checklist 3 - Step 2** : Menghitung banyaknya kemunculan label yang sesuai standart (diasumsikan 'pass' dan 'fail')

In [ ]:
# Memfilter label 'pass' yang sesuai standart ketentuan dan jumlah kemunculannya
df_pass_standart= df[df['status'] =='pass']
display(df_pass_standart)
print(f"Banyaknya data berlabel 'pass' yang sesuai standart : {df_pass_standart.shape[0]}\n")
# Memfilter label 'fail' yang sesuai standart ketentuan dan jumlah kemunculannya
df_fail_standart= df[df['status'] =='fail']
display(df_fail_standart)
print(f"Banyaknya data berlabel 'fail' yang sesuai standart : {df_fail_standart.shape[0]}\n")

,student_id,name,gender,class,ai_score,status
0,S001,Jonathan,Male,A,82.0,pass
1,S002,Aisha,Female,B,78.0,pass
2,S003,Jonas,Male,A,NaN,pass
4,S005,Liam,Male,B,88.0,pass
5,S006,Fatima,Female,A,91.0,pass
7,S008,Sofia,Female,B,110.0,pass
8,S009,Kai,Male,C,83.0,pass
12,S013,Kenichi,Male,B,80.0,pass
14,S015,Idris,Male,C,85.0,pass
16,S017,Ryan,Male,B,88.0,pass


Banyaknya data berlabel 'pass' yang sesuai standart : 10



,student_id,name,gender,class,ai_score,status
3,S004,Zara,Female,C,74.0,fail
6,S007,Ahmed,Male,C,76.0,fail
9,S010,Elena,Female,A,69.0,fail
11,S012,Amara,Female,C,72.0,fail
13,S014,Leila,Female,A,-10.0,fail
17,S018,Omar,Male,C,73.0,fail


Banyaknya data berlabel 'fail' yang sesuai standart : 6



### **Checklist 3 - Conclusion**

Apakah pelabelan status siswa yang tidak konsisten atau tidak diseragamkan akan mempengaruhi hasil keputusan sistem?
>Ya, karena pada kolom **status** ditemukan sebanyak **4 kategori** yaitu; **PASS, Pass, pass**, dan **fail**. Saya asumsikan jika kategori standar yang ditentukan adalah **pass** dan **fail** sebagai patokan kelulusan :
- Sebanyak **10 siswa** dengan status **pass**  diluluskan
- Sebanyak **6 siswa** dengan status **fail** tidak lulus
- **2 siswa** status kelulusannya tidak sesuai standart (**1 siswa** status **PASS** dan **1 siswa** status **Pass**)

Berdasarkan hasil <code><b>df_pass_standart = df[df['status'] == 'pass']</b></code> di step 2, outputnya menganggap siswa berstatus **PASS** dan **Pass** menjadi tidak lulus, karena tidak sesuai ketentuan sistem.

Jadi **2 siswa** yang seharusnya lulus akan diabaikan sistem, disebabkan oleh label kelulusannya tidak sesuai ketentuan.




## **2.5 - Solusi Checklist 4**

**Checklist 4 - Step 1** : Normalisasi format non standart agar dapat ikut dianalisis
> *Langkah ini dilakukan hanya untuk keperluan analisis berikut, bukan perbaikan terhadap sistem. Inkonsistensi label tetap merupakan temuan valid di checklist 3.*

In [ ]:
# Membuat kolom baru status_normalized untuk menyeragamkan perlabelan kelulusan
df['status_normalized'] = df['status'].str.lower()
display(df)



,student_id,name,gender,class,ai_score,status,status_normalized
0,S001,Jonathan,Male,A,82.0,pass,pass
1,S002,Aisha,Female,B,78.0,pass,pass
2,S003,Jonas,Male,A,NaN,pass,pass
3,S004,Zara,Female,C,74.0,fail,fail
4,S005,Liam,Male,B,88.0,pass,pass
5,S006,Fatima,Female,A,91.0,pass,pass
6,S007,Ahmed,Male,C,76.0,fail,fail
7,S008,Sofia,Female,B,110.0,pass,pass
8,S009,Kai,Male,C,83.0,pass,pass
9,S010,Elena,Female,A,69.0,fail,fail


**Checklist 4 - Step 2** : Mencari rata-rata 'ai_score' berdasarkan 'gender'




In [ ]:
# Mencari rata-rata 'ai_score' berdasarkan dari kolom 'gender'
display(df.groupby('gender')['ai_score'].agg(jumlah='size', rata_ai_score='mean').reset_index())


,gender,jumlah,rata_ai_score
0,Female,8,70.125000
1,Male,10,83.333333


**Checklist 4 - Step 3** : Mencari tingkat kelulusan berdasarkan persentase 'gender'.



In [ ]:
# Mencari jumlah kemunculan label kelulusan 'status_normalized' berdasarkan dari kolom 'gender'
df_status_percent = df.groupby('gender')['status_normalized'].value_counts().reset_index()

# Membuat kolom 'percentage' dari jumlah kemunculan 'count'
df_status_percent['percentage'] = (
    df_status_percent['count'] /
    df_status_percent.groupby('gender')['count'].transform('sum') * 100
)

display(df_status_percent)



,gender,status_normalized,count,percentage
0,Female,fail,4,50.0
1,Female,pass,4,50.0
2,Male,pass,8,80.0
3,Male,fail,2,20.0


### **Checklist 4 - Conclusion**

Apakah ada indikasi bias dari hasil penilaian sistem terhadap kelompok siswa berdasarkan background gender siswa? apakah rata-rata skor dan tingkat kelulusan terdistribusi secara proporsional dan adil?
> Ada indikasi bias terhadap gender siswa yang mempengaruhi kelulusan yang perlu investigasi lebih lanjut;
- Jumlah siswa gender **female** (perempuan) sebanyak **8** dibanding **male** (laki-laki) yaitu **10**, terdapat ketidakseimbangan proporsional dari dataset yang mempengaruhi aspek fairness.
- Rata-rata siswa **female** (perempuan) lebih rendah daripada siswa **male** (laki-laki), yaitu **perempuan 70.12** dibanding **laki-laki 83.33**. Perbandingan rata-rata ini juga mempengaruhi penilaian oleh sistem.
- Siswa **female** (perempuan) persentase kelulusan adalah **pass 50%** dibanding **fail 50%**, sedangkan siswa **male** (laki-laki) memiliki persentase yaitu **pass 80%** dibanding **fail 20%**. Perlu investigasi lanjutan mengenai perbedaan ratio kelulusan berdasar gender agar tidak bias sistemik.

# 3. Laporan Audit



## **3.1 - Temuan**

1. Ditemukan **3 data** yang memiliki nilai **Nan, -10,** dan **110** tetap mendapat status kelulusan, padahal nilai tersebut merupakan nilai yang diluar ketentuan nilai skala 0-100. Sistem beresiko tetap memberikan keputusan kelulusan pada nilai tidak valid karena tidak ada validasi penginputan nilai. Resiko : **High**
2. Pada kolom **status** ditemukan sebanyak **4 kategori** yaitu; **PASS, Pass, pass**, dan **fail**. Saya asumsikan jika kategori yang ditentukan adalah **pass** dan **fail** sebagai patokan kelulusan, maka siswa yang kategorinya tidak sesuai akan diabaikan oleh sistem. Resiko : **High**
3. Dataset menunjukkan keragaman nilai yang tinggi (standar deviasi **std: 24**). Investigasi per gender menunjukkan kesenjangan yang tidak merata;
>- Rata-rata nilai **female** (perempuan) sebesar **70.12**, lebih rendah daripada **male** (laki-laki) dengan rata-rata nilai **83.33**.
>- Ditemukan juga kesenjangan persentase kelulusan pada siswa **female** (perempuan) dengan tingkat kelulusan **pass 50%** dibandingkan dengan siswa **male** (laki-laki) memiliki persentase yaitu **pass 80%**.

>- Perbedaan di atas memerlukan investigasi lebih lanjut untuk menentukan apakah ini mencerminkan kemampuan siswa atau indikasi bias sistemik. Resiko : **Medium**




## **3.2 - Rekomendasi**

1. Perlu adanya validasi input sebagai ketentuan standar agar nilai di kolom 'ai_score' selalu berupa angka valid dan selalu dalam rentang skala nilai 0-100. Pengaturan ini ditujukan untuk Developer/Engineer Model untuk menindaklanjuti.
2. Standarisasi kategori di kolom status perlu diterapkan agar tidak ada data yang beresiko diabaikan oleh sistem karena tidak sesuai ketentuan meskipun siswa mempunyai kriteria yang memenuhi kelulusan. Hal ini harus diatur oleh bagian Developer/Engineer.
3. Sampling data perbandingan rata-rata per gender harus diseimbangkan baik dari representasi gender. Tingkat persentase kelulusan tiap gender juga lebih tinggi pada gender laki-laki daripada perempuan. Diperlukan Investigasi lebih lanjut untuk menentukan apakah perbedaan distribusi gender dalam dataset mencerminkan bias dalam pengumpulan data atau bias sistemik dalam model. Kepada bagian Data Engineer/Data Scientist/AI Ethic Analyst untuk mengevaluasi bias pada model.



## **3.3 - Limitasi**

1. Dataset ini hanya berisi 18 baris data dan merupakan hasil dari bantuan AI generatif dengan tujuan menjadi bahan latihan audit.
2. Analisis fairness di checklist 4 menggunakan label yang dinormalisasi, namun oleh sistem sebenarnya mengabaikan label tidak standar. Hasil persentase kelulusan per gender mungkin berbeda jika dihitung berdasarkan sistem yang sebenarnya.
3. Audit fairness hanya dilakukan berdasarkan kategori gender. Untuk kategori kelas tidak dianalisis karena jumlah sampel per kelas terlalu kecil untuk menghasilkan kesimpulan yang signifikan secara statistik.
4. Penentuan rata-rata ai_score pada tiap gender mengabaikan nilai NaN karena merupakan kalkulasi default Pandas. Jadi jumlah siswa yang dihitung per gender mungkin tidak setara.
5. Dataset ini tidak menyertakan dokumentasi tentang bagaimana ai_score dihasilkan, apakah murni output model, transformasi dari nilai asli, atau kombinasi keduanya. Kesimpulan audit perlu diverifikasi dengan dokumentasi sistem yang lengkap.


